### PR TITLE
[Core][VLM] Support image embeddings as input

### LINKED ISSUE #6604: [Feature]: MultiModal LLM with vector API
### 🚀 The feature, motivation and pitch

Consider a scenario where a large model is deployed in the cloud, and the application is deployed on a computationally limited embedded device.

If we want to support multimodal dialogue interaction with vision and language, each request would send an image (considering the dialogue history, there would be many images). Given network bandwidth and other factors, this would cause a lot of latency.

Therefore, if the VLM's image encoder and projector are deployed on the embedded device, and if we could send the encoded vector instead during requests, the data transmission volume would be much smaller. This would reduce latency and improve the user experience.

### Alternatives

The suggestted usage method is as follow

``` python
# Refer to the HuggingFace repo for the correct format to use
prompt = "USER: <vector>\nWhat is the content of this image?\nASSISTANT:"

# Image encoded vector
vector = np.array([x, x,x, x])

# Single prompt inference
outputs = llm.generate({
    "prompt": prompt,
    "multi_modal_data": {"vector": vector},
})
```

For this usage, deploying only a single-model LLM model could support multi-modal model usage, and the modality is not limited.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ywang96**
This PR adds the support for passing image embeddings as input so that they can be directly consumed by the language model.

Example usage

``` python
# Refer to the HuggingFace repo for the correct format to use
prompt = "USER: <image>\nWhat is the content of this image?\nASSISTANT:"

# Image embedding generated from a separate vision tower component, directly 
# used to be merged with text embedding.
image_embeds: torch.Tensor = ... # shape of (1, image_feature_size, hidden_size of LM)

# Single prompt inference
outputs = llm.generate({
    "prompt": prompt,
    "multi_modal_data": {"image": image_embeds},
})
```

FIXES #6604 

Follow-up TODO: Support initializing VLM with only language model backbone.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by ywang96** [comment]
@Isotr0py It looks like the generation of image embedding from pixel values and merging with text embedding is currently tied together under `Phi3HDImageEmbedding`. Could you take a look to decouple them for the effort of this PR?

File: vllm/model_executor/models/llava_next.py [hunk 1]
```diff
@@ -187,7 +193,7 @@ def input_processor_for_llava_next(ctx: InputContext, llm_inputs: LLMInputs):
             input_width=width,
         )
     elif isinstance(image_data, torch.Tensor):
-        raise NotImplementedError("Embeddings input is not supported yet")
+        return
```
Comment (DarkLight1337) [inline vllm/model_executor/models/llava_next.py:196]:
```suggestion
        return llm_inputs
```

Comment (ywang96) [reply vllm/model_executor/models/llava_next.py:196]: Good catch!

**Comment by Isotr0py** [comment]
@ywang96 Ok, I will decouple them tonight. (Sorry that I don't have bandwidth at daytime)

**Comment by ywang96** [comment]
> @ywang96 Ok, I will decouple them tonight. (Sorry that I don't have bandwidth at daytime)

No rush at all, and thank you for helping out!

**Comment by ywang96** [comment]
@DarkLight1337 Please give this PR a first pass - I have updated all vision language models except two: 
- `Chameleon` (since the model itself is only input embedding based).
- `MiniCPMV` (because the current implementation support multi-image inputs, IMO embedding support should be added after we support multi-image generally).

I plan to add a test for only Llava with embedding as input since it's not worth testing this feature on all models when the logic is the same.

On a side note, I've standardized the code organization across all files. Constants and input types will always be at the top, and the inference pipeline in the model `forward` will always follow the pattern below:

``` python
        image_input = self._parse_and_validate_image_input(**kwargs)

        if image_input is not None:
            vision_embeddings = self._process_image_input(image_input)
            inputs_embeds = self.language_model.model.get_input_embeddings(
                input_ids)

            inputs_embeds = merge_vision_embeddings(
                input_ids, inputs_embeds, vision_embeddings,
                self.config.image_token_index)

            input_ids = None
        else:
            inputs_embeds = None
```

**Review by DarkLight1337 — COMMENTED** [review]
LGTM. Just need to add a test involving embeddings input.

**Comment by DarkLight1337** [comment]
The only small change I would make is to add an `assert_never` guard at the end of each `_parse_and_validate_image_input` function to make sure that we have handled all of the cases.

**Comment by DarkLight1337** [comment]
Oh, actually - we should update the input processors to allow embeddings input. It appears that you've only updated the one for LLaVA-NeXT.

**Comment by ywang96** [comment]
> The only small change I would make is to add an `assert_never` guard at the end of each `_parse_and_validate_image_input` function to make sure that we have handled all of the cases.

I think `assert_never` is useful when we check against a particular variable through different conditions, which is a bit different from what we're doing here in `_parse_and_validate_image_input`. Maybe an `assert False` in the end?

**Comment by ywang96** [comment]
On a side note, I realized supporting image embeddings as input is also not feasible for `Fuyu` due to the image processor adding additional logics with tokenizer. Maybe @Isotr0py has more context in this?

**Comment by DarkLight1337** [comment]
> > The only small change I would make is to add an `assert_never` guard at the end of each `_parse_and_validate_image_input` function to make sure that we have handled all of the cases.
> 
> I think `assert_never` is useful when we check against a particular variable through different conditions, which is a bit different from what we're doing here in `_parse_and_validate_image_input`. Maybe an `assert False` in the end?

Maybe we can raise a `TypeError` with an informative user message in this case then.

**Comment by ywang96** [comment]
@DarkLight1337 This PR is ready for final review. I have added a test with Llava 1.5 and updated the documentation.

**Review by DarkLight1337 — APPROVED** [review]
Thanks for implementing this!

**Comment by Andcircle** [comment]
> Follow-up TODO: Support initializing VLM with only language model backbone.

@ywang96 @DarkLight1337 Thanks for the update!

I checked the demo code, we still provide 2 modality separately, prompt and images, and the merge process is still controlled by only the VLLM supported VLM model, it is not that flexible if we wanna our own merge methods.

Can we do as following: so we just take customized VLM as a PURE language model

```
#start from pure language model, NOT existing VLM
llm = LLM(model="mistral", ...)
#merge_inputs is a customized function, provided by user, this will make the process more flexible
inputs_embeds = merge_inputs(texts, images)
#the LLM only takes batch of merged embeddings, it doesn't care is it image / video / audio anymore, it just take it as pure language model
outputs = llm.generate(inputs_embeds=inputs_embeds, ...)
```

I saw you also mentioned:
Follow-up TODO: Support initializing VLM with only language model backbone.
Is this as above mentioned?

Again, really really appreciated your help.

**Comment by ywang96** [comment]
Hey @Andcircle! Thanks for reaching out!

Yes as you mentioned, what this PR does is to allow image embeddings as input so that users can process image to embeddings separately and won't need to load vision encoder and projector together with the language model in the same host (which is the TODO).

We do have a PR from a community member regarding allow embeddings as input #6869. This touches a broader aspect of the framework that we're still evaluating, but perhaps we should prioritize it given this is such a popular request. I assume if we support this feature, then you will only need to initialize the language model itself (other than initialize a VLM but only loading the language model).

Please let me know if I missed anything!

**Comment by Andcircle** [comment]
@ywang96 
Thanks for your fast response!
Yes, I think #6869 should free us.

Just to be clarified, #6869 's use case can be much broader =) not only for embedding tuning as stated, It can used for VLM research in a very flexible manner.

**Comment by ywang96** [comment]
> @ywang96 Thanks for your fast response! Yes, I think #6869 should free us.
> 
> Just to be clarified, #6869 's use case can be much broader =) not only for embedding tuning as stated, It can used for VLM research in a very flexible manner.

Yep! I totally agree with allowing embedding as input for LM is the most flexible. 

Just to be transparent, there are a few considerations on why we haven't supported it; for example, how to make prefix caching work with embeddings as input? Today we use prompt token ids as the hash/identifier of the prompt prefix, so this feature wouldn't work at all for embeddings as input.

Anyways, perhaps we could pursue for enabling it as an experimental feature and see how it goes!

**Comment by Andcircle** [comment]
> > @ywang96 Thanks for your fast response! Yes, I think #6869 should free us.
> > Just to be clarified, #6869 's use case can be much broader =) not only for embedding tuning as stated, It can used for VLM research in a very flexible manner.
> 
> Yep! I totally agree with you that allowing embedding as input for LM is the most flexible.
> 
> Just to be transparent, there are a few considerations on why we haven't supported it; for example, how to make prefix caching work with embeddings as input? Today we use prompt token ids as the hash/identifier of the prompt prefix, so this feature wouldn't work at all for embeddings as input.
> 
> Anyways, perhaps we could pursue for enabling it as an experimental feature and see how it goes!

@ywang96 
Thanks for your explanation!

Sorry, naive question, what do you mean prompt prefix? the repeated wording for each prompt?

**Comment by ywang96** [comment]
> > > @ywang96 Thanks for your fast response! Yes, I think #6869 should free us.
> > > Just to be clarified, #6869 's use case can be much broader =) not only for embedding tuning as stated, It can used for VLM research in a very flexible manner.
> > 
> > 
> > Yep! I totally agree with you that allowing embedding as input for LM is the most flexible.
> > Just to be transparent, there are a few considerations on why we haven't supported it; for example, how to make prefix caching work with embeddings as input? Today we use prompt token ids as the hash/identifier of the prompt prefix, so this feature wouldn't work at all for embeddings as input.
> > Anyways, perhaps we could pursue for enabling it as an experimental feature and see how it goes!
> 
> @ywang96 Thanks for your explanation!
> 
> Sorry, naive question, what do you mean prompt prefix? the repeated wording for each prompt?

@Andcircle Yes! For example, a long system prompt from the chat template.

**Comment by Andcircle** [comment]
@ywang96 
Will have a look, thank =)

**Comment by ywang96** [comment]
@Isotr0py Hey do you think it makes sense to support image embeddings for Fuyu? We don't have to but I just want to get an opinion from you since you added the support for it. Thanks!

**Comment by whyiug** [comment]
@ywang96 
This is a fantastic feature!
I've encountered a tricky problem.
I need to perform multiple VQA tasks (same image, different questions) using the same model architecture (e.g., Paligemma). In this case, I can use this feature to deploy image encoding and language model inference in different locations. This saves time by avoiding redundant image encoding for each language model inference.
However, how can I build an on-the-fly inference service (like a compatible server or something similar)? At this point, the solution I can think of is having the image encoding service write the encoded tensors to shared storage, while the language model inference service (vllm service) reads from the shared storage (which would involve modifying the vllm source code, building, and deploying).

**Comment by DarkLight1337** [comment]
> How can I build an on-the-fly inference service (like a compatible server or something similar)? At this point, the solution I can think of is having the image encoding service write the encoded tensors to shared storage, while the language model inference service (vllm service) reads from the shared storage (which would involve modifying the vllm source code, building, and deploying).

I suggest to modify `vllm.multimodal.utils.get_and_parse_image` and `vllm.multimodal.utils.async_get_and_parse_image` so that instead of the PIL image, it would fetch the embeddings of the image corresponding to the given URL. There should be no need to modify the rest of the code.

**Comment by whyiug** [comment]
> > How can I build an on-the-fly inference service (like a compatible server or something similar)? At this point, the solution I can think of is having the image encoding service write the encoded tensors to shared storage, while the language model inference service (vllm service) reads from the shared storage (which would involve modifying the vllm source code, building, and deploying).
> 
> From the vLLM side, I suggest to modify `vllm.multimodal.utils.get_and_parse_image` and `vllm.multimodal.utils.async_get_and_parse_image` so that instead of the PIL image, they would fetch the embeddings of the image corresponding to the given URL. There should be no need to modify the rest of the code.

That's exactly what I want to do.

**Comment by whyiug** [comment]
When I was about to add this feature to qwen2vl. Unfortunately, I've run into some difficulties.
For example, I can't just rely on image embedding to generate new prompt_token_ids without the original image.  See [here ](https://github.com/vllm-project/vllm/blob/4bb98f2190aaf408cb063df5184829fb54ee5f81/vllm/model_executor/models/qwen2_vl.py#L788.)
```python
    height, width = get_image_size(image, channel_dim=input_data_format)
```
And [here](https://github.com/vllm-project/vllm/blob/4bb98f2190aaf408cb063df5184829fb54ee5f81/vllm/model_executor/models/qwen2_vl.py#L564C5-L564C33), if we just return image embeds, it will occur an error. `AssertionError: mrope embedding type requires multi-modal input mapper returns 'image_grid_thw' or 'video_grid_thw'.` 

Might we need to passthrough more parameters for qwen2vl? please me give some tips.

**Comment by DarkLight1337** [comment]
> When I was about to add this feature to qwen2vl. Unfortunately, I've run into some difficulties. For example, I can't just rely on image embedding to generate new prompt_token_ids without the original image. See [here ](https://github.com/vllm-project/vllm/blob/4bb98f2190aaf408cb063df5184829fb54ee5f81/vllm/model_executor/models/qwen2_vl.py#L788.)
> 
> ```python
>     height, width = get_image_size(image, channel_dim=input_data_format)
> ```
> 
> And [here](https://github.com/vllm-project/vllm/blob/4bb98f2190aaf408cb063df5184829fb54ee5f81/vllm/model_executor/models/qwen2_vl.py#L564C5-L564C33), if we just return image embeds, it will occur an error. `AssertionError: mrope embedding type requires multi-modal input mapper returns 'image_grid_thw' or 'video_grid_thw'.`
> 
> Might we need to passthrough more parameters for qwen2vl? please me give some tips.

Can you open a new issue for this? (since it's specific to Qwen2-VL)

**Comment by whyiug** [comment]
> > When I was about to add this feature to qwen2vl. Unfortunately, I've run into some difficulties. For example, I can't just rely on image embedding to generate new prompt_token_ids without the original image. See [here ](https://github.com/vllm-project/vllm/blob/4bb98f2190aaf408cb063df5184829fb54ee5f81/vllm/model_executor/models/qwen2_vl.py#L788.)
> > ```python
> >     height, width = get_image_size(image, channel_dim=input_data_format)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And [here](https://github.com/vllm-project/vllm/blob/4bb98f2190aaf408cb063df5184829fb54ee5f81/vllm/model_executor/models/qwen2_vl.py#L564C5-L564C33), if we just return image embeds, it will occur an error. `AssertionError: mrope embedding type requires multi-modal input mapper returns 'image_grid_thw' or 'video_grid_thw'.`
> > Might we need to passthrough more parameters for qwen2vl? please me give some tips.
> 
> Can you open a new issue for this? (since it's specific to Qwen2-VL)
I've created a new issue and will be submitting the draft code.
https://github.com/vllm-project/vllm/issues/8857

**Pull request merged**